[MCCAS] Introduce llvm-cas-mc-dump tool

felipepiovezan opened the pull request:
This is an initial attempt at a tool that navigates the CAS tree of an
object using the MC Schema and prints the tools structure.

Currently, it only prints the node kind, followed by its ID, followed by
a nested list of children nodes (recursively). Since the CAS is always a
DAG, there is no risk of infinite loops here.

For example, one possible output:

```
mc:assembler    llvmcas://5f77...
  mc:header       llvmcas://5264...
  mc:group        llvmcas://dfa25...
    mc:section      llvmcas://1a762...
      mc:debug_info_cu llvmcas://64c13...
      mc:debug_info_cu llvmcas://d0c6c744...
```

Since the first use case for this tool is to check how many objects each debug
section contains, we provide a flag to filter non-debug sections.

There are limitations to the tool at this point, all of which can be resolved
in the future:

1. It does not understand the _contents_ of each CAS object beyond the "type"
of object it is. So, for example, we cannot print the name of a section.

2. If one object contains multiple edges with the same destination, the MC CAS
encoding may choose to only add unique edges, and encode the duplicates inside
the "Data" section of the object. The tool currently does not look at those, so
there may be situations in which duplicate edges are not printed.

3. Because we don't use any nesting marks other than whitespaces, and
FileCheck does not usually distinguish whitespaces, nesting levels may
be ambiguous for FileCheck. We should explore using `{}` in the future.

felipepiovezan committed:
[MCCAS] Introduce llvm-cas-mc-dump tool

This is an initial attempt at a tool that navigates the CAS tree of an
object using the MC Schema and prints the tree structure.

Currently, it only prints the node kind, followed by its ID, followed by
a nested list of children nodes (recursively). Since the CAS is always a
DAG, there is no risk of infinite loops here.

For example, one possible output:

```
mc:assembler    llvmcas://5f77...
  mc:header       llvmcas://5264...
  mc:group        llvmcas://dfa25...
    mc:section      llvmcas://1a762...
      mc:debug_info_cu llvmcas://64c13...
      mc:debug_info_cu llvmcas://d0c6c744...
```

Since the first use case for this tool is to check how many objects each debug
section contains, we provide a flag to filter non-debug sections.

There are limitations to the tool at this point, all of which can be resolved
in the future:

1. It does not understand the _contents_ of each CAS object beyond the "type"
of object it is. So, for example, we cannot print the name of a section.

2. If one object contains multiple edges with the same destination, the MC CAS
encoding may choose to only add unique edges, and encode the duplicates inside
the "Data" section of the object. The tool currently does not look at those, so
there may be situations in which duplicate edges are not printed.

3. Because we don't use any nesting marks other than whitespaces, and
FileCheck does not usually distinguish whitespaces, nesting levels may
be ambiguous for FileCheck. We should explore using `{}` in the future.

adrian-prantl commented on clang/test/CAS/debug_info_split.c:
```diff
@@ -2,6 +2,12 @@
+// RUN:   -Xclang -fcas-backend-mode=casid \
+// RUN:   -Xclang -fcas-path -Xclang %t.cas
+// RUN: llvm-cas-mc-dump --cas=%t.cas --casid-file %t.casid --dwarf-sections-only | FileCheck %s
+
```
The primary test for an LLVM tool shouldn't be in clang. Could we create LLVM IR out of this C file (explain in a comment how it was generated) and move the test to `llvm/test/CAS`?

felipepiovezan replied:
Yeah, I wasn't too happy about this. Let me give the IR idea a try

adrian-prantl commented on llvm/tools/llvm-cas-mc-dump/MCCASPrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+//===- llvm-cas-object-format.cpp - Tool for the CAS object format --------===//
```
This comment looks copied&pasted.

felipepiovezan replied:
Oops, it must certainly is.

adrian-prantl commented on llvm/tools/llvm-cas-mc-dump/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+  )
+
+add_llvm_tool(llvm-cas-mc-dump
+  llvm-cas-mc-dump.cpp
```
Is there a good reason not to call this a more generic `llvm-cas-dump`?

felipepiovezan replied:
It rely very heavily on this concept of "Schemas", and I am not sure if the base class interface of a Schema is generic enough for this tool. I'll check and report back

rastogishubham replied:
I second Adrian's comment, wouldn't `llvm-cas-dump` be better since we just want to dump out the cas contents?

adrian-prantl commented on llvm/tools/llvm-cas-mc-dump/MCCASPrinter.cpp:
```diff
@@ -0,0 +1,73 @@
+};
+
+bool isDwarfSection(MCObjectProxy MCObj) {
+  // Currently, the only way to detect debug section is through the kind of its
```
sections

adrian-prantl commented on llvm/tools/llvm-cas-mc-dump/MCCASPrinter.h:
```diff
@@ -0,0 +1,47 @@
+private:
+  PrinterOptions Options;
+  llvm::mccasformats::v1::MCSchema MCSchema;
+  int Indent;
```
`= 0;` ?

felipepiovezan replied:
This is initialized by the constructor, do you think we should do it here instead? AFAIK LLVM doesn't do this very often

adrian-prantl commented on llvm/tools/llvm-cas-mc-dump/llvm-cas-mc-dump.cpp:
```diff
@@ -0,0 +1,72 @@
+//===- llvm-cas-object-format.cpp - Tool for the CAS object format --------===//
```
Update?

felipepiovezan commented:
Addressed some of the comments, testing updates TBD

cachemeifyoucan reviewed (approved):
I like the idea of this tool. LGTM other than Adrian's comments.

rastogishubham reviewed (approved):
LGTM!

felipepiovezan commented:
Rebased and addressed review comments. In particular:

* Removed "mc" from the tool's name.
* Removed the clang test and created a test for the tool itself using an `ll` file

adrian-prantl commented on llvm/tools/llvm-cas-dump/MCCASPrinter.h:
```diff
@@ -0,0 +1,47 @@
+//===- llvm-cas-object-format.cpp - Tool for the CAS object format --------===//
```
Comment is out of sync

felipepiovezan replied:
My bad, I thought I had updated all of these

felipepiovezan commented:
@swift-ci please test

felipepiovezan commented:
The rebase introduced some extra "padding" nodes, so I had to update the test

felipepiovezan commented:
@swift-ci please test
